parsers/IN_HP.py

```diff
@@ -0,0 +1,163 @@
+        'production': combine_gen(get_state_gen(soup), get_isgs_gen(soup)),
+        'source': url
+    }
+
+    # TODO: Below is used for testing, remove once parser complete.
+    # with open('/tmp/hpdata.txt', 'w') as text_file:
+    #     text_file.write(res.text)
+
+    # with open('/tmp/hpdata.txt', 'r') as text_file:
+    #     soup = BeautifulSoup(text_file, 'html.parser')
+    #     return {
+    #         'zoneKey': ZONE_KEY,
+    #         'datetime': arrow.now(TZ).datetime,
+    #         'production': combine_gen(get_state_gen(soup), get_isgs_gen(soup)),
+    #         'source': url
+    #     }
+
+
+def get_state_gen(soup):
+    """Gets the total generation by type from state powerplants (MW).
+    Data is from the table titled GENERATION OF HP(Z)"""
+    table_name = 'GENERATION OF HP(Z)'
+    gen = {GenType.HYDRO.value: 0.0, GenType.UNKNOWN.value: 0.0}
+    # Read all rows except the last (Total).
+    for row in get_table_rows(soup, 'table_5', table_name)[:-1]:
+        try:
+            cols = row.find_all('td')
+            gen_type = get_gen_type(cols[0].text)
+            gen[gen_type.value] += float(cols[1].text)
+        except (AttributeError, IndexError, ValueError):
+            raise Exception(
+                'Error importing data from row: {}'.format(row))
+    return gen
+
+
+def get_isgs_gen(soup):
+    """Gets the total generation by type from ISGS powerplants (MW).
+    ISGS means Inter-State Generating Station: one owned by multiple states.
+    Data is from the table titled (B1)ISGS(HPSLDC WEB PORTAL)"""
+    table_name = '(B1)ISGS(HPSLDC WEB PORTAL)'
+    gen = {GenType.HYDRO.value: 0.0, GenType.UNKNOWN.value: 0.0}
+    # Read all rows except the first (headers) and last two (OTHERISGS and Total).
+    for row in get_table_rows(soup, 'table_4', table_name)[1:-2]:
+        try:
+            cols = row.find_all('td')
+            if not cols[0].has_attr('class'):
+                # Ignore COMPANY column.
+                del cols[0]
+            gen_type = get_gen_type(cols[0].text)
+            gen[gen_type.value] += float(cols[2].text)
+        except (AttributeError, IndexError, ValueError):
+            raise Exception(
+                'Error importing data from row: {}'.format(row))
+    return gen
+
+
+def get_table_rows(soup, container_class, table_name):
+    """Gets the table rows in the div identified by the provided class."""
+    try:
+        rows = soup.find('div', class_=container_class).find(
+            'tbody').find_all('tr')
+        if len(rows) == 0:
+            raise ValueError
+        return rows
+    except (AttributeError, ValueError):
+        raise Exception('Error reading table {}'.format(table_name))
+
+
+def get_gen_type(desc):
+    """Gets the generation type for the plant described by the given
+    description string. Raises an error if the description is not recognised."""
+    for plant_type in PLANT_NAMES_TO_TYPES.items():
+        if plant_type[0] in desc.upper():
```

Comment:
We want to be careful with this sort of partial matching, it might be better to use `==` and use the exact plant names provided (`BASPA(3x100MW)`) but I understand the full names are kind of annoying.

Also if you're using `.items()` here you can doing the following for readability.

```python
    for plant_name, plant_type in PLANT_NAMES_TO_TYPES.items():
        if plant_name in desc.upper():
            return plant_type
```

Reply:
Done, using exact matches actually allows me to simplify the code.